vllm/prompt_adapter/layers.py

```diff
@@ -19,37 +20,58 @@ class VocabParallelEmbeddingWithPromptAdapter(nn.Module):
     def __init__(self, base_layer: VocabParallelEmbedding) -> None:
         super().__init__()
         self.base_layer = base_layer
-        self.embedding_tensors: Dict[int, torch.Tensor] = {}
+        self.emb_layer = self.base_layer
+        if 'LoRA' in base_layer.__class__.__name__:
+            self.emb_layer = self.base_layer.base_layer
+
+    def create_prompt_adapter_weights(
+            self, prompt_adapter_config: PromptAdapterConfig):
+        self.embeddings_tensors = torch.zeros(
+            (
+                prompt_adapter_config.max_prompt_adapters,
+                prompt_adapter_config.max_prompt_adapter_token,
+                self.emb_layer.embedding_dim,
+            ),
+            dtype=self.emb_layer.weight.dtype,
+            device=self.emb_layer.weight.device,
+        )
+        self.adapter_lengths = torch.zeros(
+            prompt_adapter_config.max_prompt_adapters,
+            dtype=torch.long,
+            device=self.emb_layer.weight.device)
         self.indices_gpu: torch.Tensor
         self.flag: bool = False
 
     def reset_prompt_adapter(self, index: int):
-        self.embedding_tensors[index] = 0
+        self.embeddings_tensors[index] = 0
 
     def set_prompt_adapter(
         self,
         index: int,
-        embeddings_tensor: Optional[torch.Tensor],
+        adapter_model: Optional[torch.Tensor],
     ):
         self.reset_prompt_adapter(index)
-        if embeddings_tensor is not None:
-            self.embedding_tensors[index] = embeddings_tensor
+        if adapter_model is not None:
+            length = adapter_model.shape[0]
+            self.embeddings_tensors[index, :length] = adapter_model
+            self.adapter_lengths[index] = length
 
     def set_mapping(
         self,
-        base_indices: List[int],
+        base_indices: torch.Tensor,
     ):
-        self.indices_gpu = torch.tensor(base_indices, device="cuda")
-        self.flag = torch.sum(self.indices_gpu) > 0
+        self.indices_gpu = base_indices.to(device=self.emb_layer.weight.device)
+        self.flag = (torch.sum(self.indices_gpu) / self.indices_gpu.shape[0] !=
+                     -1).item()
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         hidden_states = self.base_layer(x)
         if self.flag:
             unique_indices = torch.unique(self.indices_gpu)
             for idx in unique_indices:
-                if idx != 0:
-                    pa_idx = self.embedding_tensors[
-                        idx.item()].prompt_embedding
+                if idx != -1:
```

Comment:
Can you explain to me the general logic of applying those adapters, especially when can we expect the idx to be -1 or non-unique?

Reply:
Yeah we need to get to a point where forward pass has no:
* CPU<->GPU synchronization (no item calls, indexing with python lists etc.)
* No python control flow

maybe we can write a simple triton kernel for this? the logic seems quite straightforward...

Reply:
Hey @Yard1, I have changed the logic to remove the loops and item calls, now it is only limited to one if statement called torch.numel which I think is CUDA graph safe.
Don't have a ton of experience in Triton, would need some time to figure that out if this isn't acceptable

Reply:
Hey @Yard1, I have simplified the logic to remove the loops and item calls, now it is only limited to one if statement called torch.numel which I think is CUDA graph safe.
Don't have a ton of experience in Triton, would need some time to figure that out if this isn't acceptable

Reply:
We have to have absolutely 0 if statements :(

Let me see if I can write a quick triton kernel for this.

Reply:
@Yard1 do you mean this is needed for CUDA graphs to work? or just that it would result in a graph break which is sub-optimal?

Reply:
@Yard1 I see `if full_output.ndim == 3:` in the forward of `VocabParallelEmbeddingWithLoRA`, is that CUDA graph safe? We can replace the torch.numel with this if this is acceptable.

Reply:
@njhill we can't have any cuda graph breaks since we capture the entire forward pass. this will mean the logic will be incorrect.

@SwapnilDreams100 any CPU control flow is not going to work (what will happen is that the cuda graph will be recorded going down one branch, and then when you replay that control flow logic will not be present in there, so you won't be able to dynamically change anything during runtime)